radicle-keystore/src/file.rs

```diff
@@ -0,0 +1,158 @@
+use std::{
+    fmt::{self, Debug, Display},
+    fs::File,
+    io,
+    marker::PhantomData,
+    path::{Path, PathBuf},
+};
+
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+use crate::{
+    crypto::{self, SealedKey},
+    HasMetadata,
+    IntoSecretKey,
+    Keypair,
+    Pinentry,
+    Storage,
+};
+
+pub struct FileStorage<P, PK, SK, M> {
+    key_file_path: PathBuf,
+    pinentry: P,
+
+    _marker: PhantomData<(PK, SK, M)>,
+}
+
+impl<P, PK, SK, M> FileStorage<P, PK, SK, M> {
+    pub fn new(key_file_path: &Path, pinentry: P) -> Self {
+        Self {
+            key_file_path: key_file_path.to_path_buf(),
+            pinentry,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn key_file_path(&self) -> &Path {
+        self.key_file_path.as_path()
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct Stored<PK, M> {
+    public_key: PK,
+    secret_key: SealedKey,
+    metadata: M,
+}
+
+#[derive(Debug)]
+pub enum Error<P> {
+    KeyExists,
+    NoSuchKey,
+    Crypto(crypto::UnsealError),
+    Pinentry(P),
+    Serde(serde_cbor::error::Error),
+    Io(io::Error),
+}
+
+impl<P: Display + Debug> std::error::Error for Error<P> {}
+
+impl<P: Display> Display for Error<P> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::KeyExists => f.write_str("Key exists, refusing to overwrite"),
+            Self::NoSuchKey => f.write_str("No key found"),
+            Self::Crypto(e) => write!(f, "Error unsealing key: {}", e),
+            Self::Pinentry(e) => write!(f, "{}", e),
+            Self::Serde(e) => write!(f, "{}", e),
+            Self::Io(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+impl<P> From<io::Error> for Error<P> {
+    fn from(e: io::Error) -> Self {
+        Self::Io(e)
+    }
+}
+
+impl<P> From<serde_cbor::error::Error> for Error<P> {
+    fn from(e: serde_cbor::error::Error) -> Self {
+        Self::Serde(e)
+    }
+}
+
+impl<P, PK, SK, M> Storage for FileStorage<P, PK, SK, M>
+where
+    P: Pinentry,
+    P::Error: Display + Debug,
+    SK: AsRef<[u8]> + IntoSecretKey<M> + HasMetadata<M>,
+    PK: Clone + From<SK> + Serialize + DeserializeOwned,
+    M: Clone + Serialize + DeserializeOwned,
+{
+    type Pinentry = P;
+
+    type PublicKey = PK;
+    type SecretKey = SK;
+
+    type Metadata = M;
+
+    type Error = Error<P::Error>;
+
+    fn put_key(&mut self, key: Self::SecretKey) -> Result<(), Self::Error> {
+        if self.key_file_path().exists() {
```

Comment:
I wonder if there's anything to be said about the fact that the setup of `put_key` is generally the same. There's some existence test, followed by the same storage logic, and ending in how to "persist" the data. This could potentially be abstracted into higher-order functions, but it could also be viewed as over-engineering :upside_down_face: Food for thought

Reply:
Same goes for `get_key` and `show_key` as well I suppose

Reply:
I'm not sure I want to argue with `rustc` about higher-order functions :upside_down_face: 

I think with the `Crypto` factored out, it doesn't look so bad: each impl just deals with the details of how things are stored, and additional serialisation where needed.

Reply:
lol, sure thang! I just missed Haskell for a moment y'know..